### PR TITLE
feat: document new anomaly detection syntax

### PR DESCRIPTION
**Pull request by baturayo**
- document new anomaly detection syntax
- lint
- Resolves: https://sodadata.atlassian.net/browse/IA-506

### CODE REVIEW

File: soda-cl/anomaly-detection.md [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+- **Learning Patterns**: At its heart, the algorithm is designed to understand and learn from the historical patterns in your data. This includes recognizing typical trends and seasonal variations in your metrics.
+
+- **Identifying Anomalies**: Once the algorithm has learned the normal behavior of your metrics, it becomes capable of detecting when something deviates from this norm. These deviations are flagged as anomalies.
+
+- **Use Case**: For instance, if your data typically shows a certain trend or pattern over the course of a week or a month, the algorithm will learn this. Then, if there's a sudden and unusual change in this pattern, it's identified as an anomaly. This could signal an issue in your data quality that requires attention.
+
+By using Facebook Prophet, this anomaly detection check offers a robust way to monitor your data quality metrics, ensuring that any irregularities are caught promptly and accurately. Another benefit of Facebook Prophet is that it's designed to be easy to use and tune, so you don't need to be an expert in machine learning to take advantage of this feature.
+
+<!---
```
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md:83]: Can I add plotly graphs here?

File: soda-cl/anomaly-detection.md [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+## Reset anomaly history
+
+If you wish, you can reset an anomaly detection's history, effectively recalibrating what Soda considers anomalous on a dataset.
+
+1. In Soda Cloud, navigate to the **Check History** page of the anomaly check you wish to reset.
+2. Click to select a node in the graph that represents a measurement, then click **Feedback**.
+3. In the modal that appears, you can choose to exclude the individual measurement, or all previous data up to that measurement, the latter of which resets the anomaly detection's history.
+
+![reset-anomaly-detection](/assets/images/reset-anomaly-detection.png){:height="600px" width="600px"}
```
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md:173]: good point, i need to change the URL to anomaly-score ,thanks

File: soda-cl/anomaly-detection.md [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+**window_length**: It sets the size of the rolling window used for training. It uses the last `window_length` data points for model training, ensuring that the model is always updated with the most recent data. The default length is `1000`. For instance, if your frequency is hourly (H), the model trains on the last 1000 hours of data.
+
+**aggregation_function**: This parameter defines how the data within each window is aggregated. The default `aggregation_function` is last, but there are several other options:
+
+- `last`: Uses the last non-null value in the window.
+- `first`: Uses the first non-null value.
+- `mean`: Calculates the average.
+- `min`: Finds the minimum value.
+- `max`: Finds the maximum value.
+- `quantile`: Calculates a specified quantile (e.g., 0.25, 0.5, 0.75).
```
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md:258]: good point, I'll test and update this field

File: soda-cl/anomaly-detection.md [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    holidays = None
+    holidays_prior_scale = 10.0
+    mcmc_samples = 0
+    uncertainty_samples = 1000
+    stan_backend = None
+    scaling = "absmax"
+    holidays_mode = None
+  ```
+
+- **mape**: [Mean absolute percentage error (MAPE)](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error) is a statistical measure of how accurate a forecasting method is. It calculates the average percentage error between the forecasted and the actual values. The lower the `MAPE` value, the more accurate the model's predictions are. When optimizing for `MAPE`, the model becomes more sensitive to changepoints and seasonal variations, providing a tighter fit to the training data. However, this increased sensitivity can sometimes lead to overfitting. In such cases, the model might mistakenly identify normal data points as anomalies. If overfitting becomes an issue, switching to `coverage` profile might be more beneficial. For your guidance, specific hyperparameters are associated with the `MAPE` as shown below:
```
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md:309]: Yes, I found the best hyperparameters for smape but there wasn't a significant difference between smape and mape. I thought `mape` is more known and that's why I prefer to use it here. If you want I can change it to smape?

File: soda-cl/anomaly-detection.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (janet-can) [reply soda-cl/anomaly-detection.md:83]: I haven't done so before, and I'm not sure. You can add images, though.

File: soda-cl/anomaly-detection.md [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+This example sets up an anomaly detection check for the count of invalid values in the `user_email` column of a `dim_customer` dataset. The `invalid_values` parameter is used to define which values are considered invalid.
+
+```yaml
+checks for dim_customer:
+  - anomaly detection for invalid_count(user_email):
+      valid format: email
+```
+
+## Anomaly detection check results
+<!--Linked to UI, access Shlink-->
```
[the remark this answers is not in the document]
Comment (janet-can) [reply soda-cl/anomaly-detection.md:139]: That's my tag to indicate that anything I change with the subheading will affect a short-link connection to the UI. Feel free to delete!

**Comment by janet-can** [comment]
@baturayo, three questions for you: 
Will anomaly detection replace anomaly score when someone adds automated checks to a data source? 

When anomaly detection is released, do we have an idea of how long anomaly score will work until it is fully deprecated? 

Is it OK with you if I use this PR to also change all other references to anomaly score in the docs to anomaly detection?

**Comment by janet-can** [comment]
@baturayo and @bastienboutonnet I've adjusted a fair bit of content in the name of the style guide and to present info a bit differently using tables in a consistent way, and such. A couple of thing I must ask of you:
1. Can you verify that the value for profile should be `mape` or `MAPE`? I found both versions used in the doc and I wasn't sure which was accurate or if it mattered. For now, I changed all instances to the allcaps version.
2. Some vales in the code snippets have double-quotes or single-quotes that I'm not sure are necessary. Can you confirm? FYI: I removed the double-quotes around the value for the `name` of the anomaly detection checks. 

The content is terrific, BTW! It is so detailed and thorough and explain a lot of complex info. Thanks ever so much for putting in the effort to make these docs great!

**Comment by janet-can** [comment]
One more thing:  
A nice-to-have in these docs would be examples that directly address the kinds of things customers have been asking of the anomaly check. Something like, "Here are the settings you can adjust to get the algorithm to discount expected abnormal values for Black Friday sales."

**Comment by baturayo** [comment]
@janet-can here are my answers. Sorry for the delay

> Will anomaly detection replace anomaly score when someone adds automated checks to a data source?
Not in the first release I guess. For me, it's safer to replace it in the second release. We can experiment a bit after the first release and collect some user feedback. Then, we can make the switch. That's my opinion. Wdyt @bastienboutonnet ?

> Will check suggestions add anomaly detection checks instead of anomaly score checks?
same answer with above

> When anomaly detection is released, do we have an idea of how long anomaly score will work until it is fully deprecated?
No idea, we need to provide quite a bit of time to make sure that all users switch to anomaly detection from anomaly score. Leaving the answer to @bastienboutonnet . 

> Is it OK with you if I use this PR to also change all other references to anomaly score in the docs to anomaly detection?
Ofc, I saw that you already did. In practice, the automated checks will still use anomaly score. We didn't implement the transition yet. So, same answer with the 1st q.

**Comment by baturayo** [comment]
> One more thing: A nice-to-have in these docs would be examples that directly address the kinds of things customers have been asking of the anomaly check. Something like, "Here are the settings you can adjust to get the algorithm to discount expected abnormal values for Black Friday sales."

I just added a part with what-if scenarios under `## Handling Common Anomaly Detection Issues` title. We will cover the holidays or special days in the next release.

File: soda-cl/anomaly-detection.md [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+```
+
+In our anomaly detection model, we have focused on fine-tuning key hyperparameters to optimize performance. Specifically, our tuning targets two main objectives: `coverage` and `mape (Mean Absolute Percentage Error)`. These objectives guide our adjustments to two critical hyperparameters: `changepoint_prior_scale` and `seasonality_prior_scale`. For detailed insights into the most effective combinations of these hyperparameters, refer to the best practices outlined [here](###Best-Practices-for-Prophet-Model)
+
+- **coverage**: This is the default recommended value. Coverage refers to the percentage of actual data points that fall within the model's predicted confidence intervals. When a prediction lies outside these intervals, it's flagged as an anomaly. A higher coverage indicates that the model is less sensitive to anomalies, making it more resilient to noise in your data. However, this might lead to underfitting, potentially causing the model to miss some anomalies. In such situations, optimizing for `MAPE` (Mean Absolute Percentage Error) might be more appropriate. For your reference, specific hyperparameters are set for the coverage profile as shown below.
```
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md]: made an update for coverage

**Comment by janet-can** [comment]
Thanks for the responses, @baturayo. The new content for example adjustments is GOLD. Our customers are going to love it; well done!  I edited a few things here and there and added to the in-page TOC. I also added a "tag" to the very top of the page to say that this check is not supported by Soda Core, as I think that's the case since I didn't see a corresponding PR in the soda-core code base. Please correct me if that assumption is wrong.
Meanwhile, if you have nothing more to add or change, it LGTM!

**Comment by bastienboutonnet** [comment]
@janet-can Until a clear decision to deprecate functionality from core, I would say this functionality will also make it to core. It's just that it's a pain in the ass and we have to do separate PRs and we started with library first. 

@baturayo will likely open a PR in core soon after initial work is merged and validated in library.

That being said, you can see add a tag pointing to the version(s) from which the new algo is going to be supported

**Comment by bastienboutonnet** [comment]
image
@janet-can @baturayo what do you think?

**Comment by bastienboutonnet** [comment]
@baturayo 
image

**Comment by bastienboutonnet** [comment]
@baturayo 
image

**Comment by bastienboutonnet** [comment]
@baturayo 
image

**Comment by bastienboutonnet** [comment]
@baturayo 
image

**Comment by janet-can** [comment]
> @janet-can Until a clear decision to deprecate functionality from core, I would say this functionality will also make it to core. It's just that it's a pain in the ass and we have to do separate PRs and we started with library first.
> 
> @baturayo will likely open a PR in core soon after initial work is merged and validated in library.
OK, for now I just put that it's not available in Soda Core yet. (It will give me a leg to stand on when it comes up in the community.)
> That being said, you can see add a tag pointing to the version(s) from which the new algo is going to be supported
Done! I assumed soda-library 1.2.2 and soda-agent 0.8.54, but can adjust as needed when the time comes.

**Comment by janet-can** [comment]
@bastienboutonnet Many of the "underfit" and "overfit" labels were my doing; I can adjust. I was thinking they were known terms to the folks brave enough to wade into Algorithm Land.

**Comment by baturayo** [comment]
> @baturayo image

good point, it's better for me to find a new graph actually or I'll play with the values a bit.

**Comment by baturayo** [comment]
> @baturayo image

Indeed MAPE was not a good option. Decreasing the window length was the only solution due to the significant pattern change. I'll try to find another graph to clearly explain.

**Comment by baturayo** [comment]
> @baturayo image

I wanted to show two things in this graph on purpose. The reason is that converting to MAPE might fix one of the problems (underfit) in the training dataset, however due to the large interval we also need to remove the anomalies from the graph. If we use the coverage profile, it still gives us large intervals due to underfit.

File: soda-cl/anomaly-detection.md [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+
+| Configuration key | Value                           | Default   |
+| ----------------- | ------------------------------- | --------- |
+| `type`            | `prophet`                       | `prophet` |
+| `profile`         | `coverage` <br /> `MAPE`        | `coverage`|
+
+
+For each of these values, Soda has adjusted the values of a few of the model's hyperparameters to tailor its performance, particularly with regard to the `changepoint_prior_scale` and `seasonality_prior_scale`. 
+
+`coverage` represents the percentage of actual measurements within the model's predicted confidence intervals. For example, if a model forecasts sales between 10-20 units on certain days and 90 out of 100 actual sales figures fall within this range, the coverage is 90%. When predictions fall outside these intervals, as in the remaining 10 cases, the model flags them as anomalies. Coverage-optimized models are more tolerant of small noises in the data that can lead to falsely-identified anomalies. However, it might underfit the data if there is a fluctuating pattern.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:260]: 1. The line 

"When predictions fall outside these intervals, as in the remaining 10 cases, the model flags them as anomalies"

feels a bit out of place here, as it seems to touch on a dynamic of our algorithm that's not necessarily related to coverage. We always use CI to flag datapoints as anomalies right?

2. Do you think we can add a link that explains coverage like we do below for MAPE?

3. “Coverage-optimized models are more tolerant of small noises in the data that can lead to falsely-identified anomalies. However, it might underfit the data if there is a fluctuating pattern.” I would rephrase this to explain that the coverage **profile** is less sensitive, as I don't see that we can say that models that are optimised for coverage are generally more sensitive.

File: soda-cl/anomaly-detection.md [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+holidays = None
+holidays_prior_scale = 10.0
+mcmc_samples = 0
+uncertainty_samples = 1000
+stan_backend = None
+scaling = "absmax"
+holidays_mode = None
+```
+
+`coverage` is less sensitive to anomalies than `MAPE`. If you have set the profile value to `coverage` and find that the model is underfitting, meaning it seems to miss some anomalies, try changing the value to `MAPE`. Conversely, if you set the value to `MAPE` and find that the model is overfitting, meaning it mistakenly identifies normal measurements as anomalies, try changing the value to `coverage`. See [Best practices for model configuration](#best-practices-for-model-configurations) for further guidance.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:314]: Not sure the words 'overfit' or 'underfit' can be used in this way. I would use sensitive and less sensitive. 

Overall comment on this section would be to maybe explain more clearly that we did experiments and found that our coverage **profile** is generally less sensitive that the mape **profile**

File: soda-cl/anomaly-detection.md [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+{% include code-header.html %}
+```yaml
+checks for dim_customer:
+  - anomaly detection for row_count:
+      name: Anomaly detection for row_count
+      model:
+        type: prophet
+        hyperparameters:
+          static:
+            profile:
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:333]: Aren't we missing the configuration that defines which profile is used?

File: soda-cl/anomaly-detection.md [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+              custom_hyperparameters:
+                seasonality_mode: additive
+                interval_width: 0.8
+                ...
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:337]: What is the reasoning behind not showing the other parameters that can be tuned? I'm a bit confused by the sentence "... not shown are the remaining parameters set to mimic the `coverage` profile settings".

File: soda-cl/anomaly-detection.md [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+```
+
+<br />
+
+
+## Add optional dynamic hyperparameter tuning configurations
+
+To dynamically tune Prophet to evaluate and select the best hyperparameters values to use before each scan, you can add a `dynamic` parameter and any number of optional hyperparameter configurations. Be aware that hyperparameter tuning can be time-consuming and resource-intensive, so best practice dictates that you use these configurations sparingly.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:345]: 1. Maybe it would make sense to more clearly explain that setting this to dynamic means that every x runs we’re going to do a grid search for the best hyperparameters?

2. "Be aware that hyperparameter tuning can be time-consuming and resource-intensive, so best practice dictates that you use these configurations sparingly."

I understand what you mean but would reformulate this part. If someone wants to use more resources to try out more configurations that’s not against best practices. Maybe wouldn’t say using resources sparingly but just warn users that testing out many different configurations can lead to longer runtimes.

File: soda-cl/anomaly-detection.md [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+```
+
+| Configuration key | Value                           | Default | 
+| ----------------- | ------------------------------- | ------- |
+| `objective_metric`| `coverage` <br /> `MSE` <br /> `RMSE`<br /> `MAE`<br /> `MAPE`<br /> `MDAPE`<br /> `SMAPE` | n/a |
+| `parallel`        | `true`<br /> `false` |  `true` |
+| `cross_validation_folds`| integer | `5` |
+| `parameter_grid`  | any Prophet-supported hyperparameters | `changepoint_prior_scale: [0.001, 0.01, 0.1, 0.5]` <br /> `seasonality_prior_scale: [0.01, 0.1, 1.0, 10.0]` <br /> other hyperparameters set to the defaults in the `coverage` profile |
+
+The `objective_metric` hyperparameter evaluates the model's performance. You can set the value to use a single string, or a list of strings. If you provide a list, the model optimizes each metric in sequence. In the example above, the model first optimizes for `coverage`, then `SMAPE` in the event of a tie. Best practice dictates that you use `coverage` as the first objective metric, and `SMAPE` as the second objective metric to optimize for a model that is more tolerant of noise in your data.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:372]: "Best practice dictates that you use `coverage` as the first objective metric, and `SMAPE` as the second objective metric to optimize for a model that is more tolerant of noise in your data." Would it make sense to say we _recommend_ first using coverage instead of using _best practice_. Best practice is quite strong and this setting the choice really just depends on the user's use case. Follow up question on this: is it true that just using coverage leads to a less sensitive model? Or did we just find that the profile that we used with coverage is less sensitive?

File: soda-cl/anomaly-detection.md [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+
+The `parallel` hyperparameter specifies whether the model saves time by using multiprocess to parallelize the cross validations.  Set the value to `True` if you have multiple cores.
+
+The `cross_validation_folds` hyperparameter sets the number of periods for each cross-validation fold. For example, with the `frequency` set to daily `D` and a `cross_validation_folds` of `5`, the model conducts cross-validation in five-day intervals. It trains on the first `n-5` days, then tests on the `n-4`th day. Subsequently, it trains on `n-4` days, testing on the `n-3`rd day, and so on. The cross-validation process computes the `objective_metric` across different data segments for each hyperparameter combination. The model then uses the best `objective_metric` according to the value or list of values configured for that hyperparameter.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:376]: I would change / rephrase this it’s not clear what you mean here to me. What do you mean with ’n’?

File: soda-cl/anomaly-detection.md [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+
+The `parameter_grid` hyperparameter is a dictionary that lists hyperparameters and their possible values. The model tests every possible combination of the listed values for each hyperparameter to identify the best value to use to detect anomalies.  You can configure any <a href="https://facebook.github.io/prophet/docs/diagnostics.html#hyperparameter-tuning:~:text=Parameters%20that%20can%20be%20tuned" target="_blank">Prophet-supported hyperparameter</a>.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:378]: Closing comment on this section; I would make it much clearer what is happening here and when it runs exactly. I would also more clearly explain what `hyperparameter_grid` means. This section contains no mention of hyperparameter grid search even though that's what we're doing? You do explain it very briefly in the next section but that feels a bit too late as it should be part the explanation of how the configuration works IMO.

File: soda-cl/anomaly-detection.md [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+
+### Execution time analysis for dynamic hyperparameter tuning
+
+The execution time for dynamic hyperparameter tuning varies based on several factors including the number of hyperparameters and the number of folds. For example, the default hyperparameter grid has 16 combinations since `changepoint_prior_scale` and `seasonality_prior_scale` have four values each. Best practice recommends that you use a small number of hyperparameters to avoid long execution times. By default, the model processes each fold in parallel. If you use multiple cores, you can set the `parallel` parameter to `True` to speed up the execution time.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:382]: Would it make sense to change the wording of best practice to something else? You use it quite a lot but I’m not sure I agree that it’s the right wording in this context. It really depends on the customer’s use case whether to use a small number of hyperparameters.

File: soda-cl/anomaly-detection.md [hunk 16]
```diff
@@ -0,0 +1,548 @@
...
+* Only use the `custom_hyperparameters` configuration if you know how Facebook Prophet works. Before making any customizations, consult the <a href="https://facebook.github.io/prophet/docs/diagnostics.html#hyperparameter-tuning:~:text=Parameters%20that%20can%20be%20tuned" target="_blank">Facebook Prophet documentation</a>. The`change_point_prior_scale` and `seasonality_prior_scale` hyperparameters have the most impact on the model so best practice dictates that you experiment with the values of these two hyperparameters first before customizing or tuning others.
+* Adjust the value of the `interval_width` hyperparameter to obtain a more anomaly-sensitive model. The default value for this hyperparameter is `0.999` which means that the model applies a confidence interval of 99.9% which, in turn, means that if the predicted value is outside of the 99.9% interval, Soda flags it as an anomaly. If you want to have a more sensitive model, you can decrease this value though be aware that a lower value may result in more falsely-identified anomalies.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:410]: Is this true? Aren’t we using / going to use our own custom confidence intervals?

File: soda-cl/anomaly-detection.md [hunk 17]
```diff
@@ -0,0 +1,548 @@
...
+
+## Address common anomaly detection issues
+
+What follows are some examples of how to adjust optional configurations to address common issues with the sensitivity of anomaly detection checks.
+
+### Insensitive detection
+
+The default `coverage` hyperparameter profile is more tolerant of small noises in data quality measurements. However, as in the following example, the profile may not be sensitive enough if there are fluctuating data patterns. The reason is that `coverage` profile uses a low `changepoint_prior_scale=0.001` value and a low `seasonality_prior_scale=0.01` which make the model less sensitive to changepoints which makes it more likely that the model misses some anomalies, as indicated in the following graph.
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:419]: Would it make sense to add more links to support you explanations? For example with links to wikipedia / fb prophet docs that explain how the change point or seasonality prior impact the behaviour of the model?

File: soda-cl/anomaly-detection.md [hunk 18]
```diff
@@ -0,0 +1,548 @@
...
+Having set the profile to `MAPE`, the graph indicates that the model is more sensitive to anomalies in data earlier in the year.
+![ad-mape-anomaly-feedbacks](/assets/images/ad-mape-anomaly-feedbacks.png){:height="700px" width="700px"}
+
+To address the second issue, use Soda Cloud to ignore the anomalies in the red rectangle. Where they crete very large anomaly boundaries, best practice dictates that you ignore the anomalies by using the **Feedback** feature. Hover over the anomalous measurement in your anomaly detection check page, then click the **Feedback** button and choose to **Ignore this value in future anomaly detection** as in the screenshot below.
+
+![ad-cloud-feedback.png](/assets/images/ad-cloud-feedback.png){:height="700px" width="700px"}
+
+After instructing Soda to ignore the anomalous measurements, the graph below indicates that the model's confidence interval is smaller and the model is more sensitive to anomalies.
+
+![ad-feedback-processed-mape](/assets/images/ad-feedback-processed-mape.png){:height="700px" width="700px"}
```
Comment (tituskx) [inline soda-cl/anomaly-detection.md:505]: Could you maybe elaborate what you mean here? The way this is phrased it comes across to me as if this sentence is related to the previous story about feedback, but I don’t see this graph anywhere else.

**Review by tituskx — COMMENTED** [review]
Left a few comments but awesome job!! Really love the new functionality; especially with the simulation this is going to be huge! Also really like the pictures of the different graphs and explanations:)

File: soda-cl/anomaly-detection.md [hunk 19]
```diff
@@ -0,0 +1,548 @@
...
+&nbsp;&nbsp;&nbsp;&nbsp;[Overly-sensitive detection](#overly-sensitive-detection)<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;[Large boundaries that ignore anomalies](#large-boundaries-that-ignore-anomalies)<br/>
+[Track anomalies and relative changes by group](#track-anomalies-and-relative-changes-by-group)<br />
+[Troubleshoot Soda Scientific installation](#troubleshoot-soda-scientific-installation)<br />
+[Go further](#go-further) <br />
+<br />
+
+## About anomaly detection checks
+
+The anomaly detection check is powered by a machine learning algorithm that works with measured values for a metric that occurs over time. Soda levergages the <a href="https://facebook.github.io/prophet/" target="_blank">Facebook Prophet</a> algorithm to learn the patterns of your data to identify and flag anomalies in your data. As a relatively easy algorithm to use and tune, Facebook Prophet is ideally suited to both analyzing metrics and giving you control over optional configurations.
```
[the remark this answers is not in the document]
Comment (janet-can) [reply soda-cl/anomaly-detection.md:79]: Grrrr. I did. Google Docs is not as reliable as I thought it was.

File: soda-cl/anomaly-detection.md [hunk 20]
```diff
@@ -0,0 +1,548 @@
...
+    - dim_customer
+  checks:
+    - anomaly detection for row_count
+```
+
+<br />
+
+## Add optional training dataset configurations
+
+A training dataset is one that Soda uses to teach the algorithm to identify patterns in the measurements the check collects. To enhance the flexibility of anomaly detection, you can add an optional `training_dataset_parameters` configuration to your anomaly detection check to customize the way that the check uses the training dataset. Apply training dataset configurations to the training dataset, time-series prediction model, and/or the anomaly detection check itself.
```
[the remark this answers is not in the document]
Comment (janet-can) [reply soda-cl/anomaly-detection.md:203]: For the first, sure.

For the second, maybe. At a certain point, the information becomes too lengthy and/or dense to absorb. If we felt examples were needed, I would suggest adding them to an examples section further down. The page is already pretty dense, in fact. If we were to add much more to it, I would seriously start thinking about breaking the page up into: this is how you use it on one page, this is all the explanation on another page. (It may have already reached this point.)

File: soda-cl/anomaly-detection.md [hunk 21]
```diff
@@ -0,0 +1,548 @@
...
+            profile: coverage 
+```
+
+| Configuration key | Value                           | Default   |
+| ----------------- | ------------------------------- | --------- |
+| `type`            | `prophet`                       | `prophet` |
+| `profile`         | `coverage` <br /> `MAPE`        | `coverage`|
+
+
+For each of these values, Soda has adjusted the values of a few of the model's hyperparameters to tailor its performance, particularly with regard to the `changepoint_prior_scale` and `seasonality_prior_scale`.
```
[the remark this answers is not in the document]
Comment (janet-can) [reply soda-cl/anomaly-detection.md:258]: We can-- what is confusing about it? (As a n00b, I didn't get lost.)

File: soda-cl/anomaly-detection.md (hunk 7 above)
Comment (janet-can) [reply soda-cl/anomaly-detection.md:260]: 1. Not for me to say! You can adjust if you wish.
2. I can, sure. I did a quick search but didn't find an explanation that I loved or was sure was 100% right. Do you have a suggestion for a learn more link?
3. Sure, I adjusted it!

File: soda-cl/anomaly-detection.md (hunk 8 above)
Comment (janet-can) [reply soda-cl/anomaly-detection.md:314]: It seems underfit and overfit are words that don't really work in the context of this doc. I'm going to replace them.

File: soda-cl/anomaly-detection.md (hunk 14 above)
Comment (janet-can) [reply soda-cl/anomaly-detection.md:378]: Maybe. I would argue that this content focuses on how to tweak the algorithm in the context of using it with this Soda check. It should not include exhaustive information about Facebook Prophet stuff, like cross validation. There is no reason we couldn't link to it to offer more info, but I would not attempt to explain it much further here as it deviates too far from the purpose of this doc. 

Secondly, IMO the explanations offer just enough info for someone to be able to self-identify as a user who ought to be tweaking these values. If you read the brief explanation and are comfortable moving ahead, maybe while referencing Facebook for a few details, then this is enough for you. If you read it and don't fully understand everything about cross validation (for example), then you ought to back away slowly. The lack of depth sends a signal to the reader that we're not here to teach you about the nuances of model functionality, we're here to tell you you can change a few things if you know what you are doing.

To address these concerns, I would suggest adding a few links to FB Prophet docs at most.

File: soda-cl/anomaly-detection.md (hunk 15 above)
Comment (janet-can) [reply soda-cl/anomaly-detection.md:382]: I use "Best practice recommends..." to avoid introducing Soda or "we" as a character in tech docs. (In extreme cases, where something goes wrong, it can be used against Soda legally as something the company specifically recommended.)

But if it's really not a best practice, then I can adjust to read more like, "Consider using a small number of hyperparameters..."

File: soda-cl/anomaly-detection.md (hunk 17 above)
Comment (janet-can) [reply soda-cl/anomaly-detection.md:419]: Perhaps. Though, again, I'm not sure I would go too deep here. The language and the topic of the content should help users self-identify about how deep they want to wade into what is clearly a complex tool. Moreoever, we've linked to Facebook Prophet a couple of times, there's no reason someone couldn't Google the terms to learn more if needed. Though one or two links wouldn't hurt.

Also, you don't *really* need to understand what those terms mean in order to tweak them. The values rather explain what is more sensitive and what is less sensitive. A user can adjust in the direction they want and/or read more about the hyperparameters in detail elsewhere.

File: soda-cl/anomaly-detection.md (hunk 18 above)
Comment (janet-can) [reply soda-cl/anomaly-detection.md:505]: Yep! Have done. 👍

File: soda-cl/anomaly-detection.md [hunk 22]
```diff
@@ -0,0 +1,427 @@
...
+
+3. **Use different `interval_width` hyperparameter if you want more sensitive model**. The default value is `0.999` which means that the model will use 99.9% confidence interval. It means that if the predicted value is outside of 99.9% interval, it will be marked as an anomaly. If you want to have a more sensitive model, you can try to decrease this value. However, it may lead to false positive alarms.
+
+4. **Use `tune` configuration only if it is necessary.** Hyperparameter tuning is an expensive process since the model tries all possible combinations of the hyperparameters. Therefore, we discourage using hyperparameter tuning unless it is really necessary. If you need to use hyperparameter tuning, then we recommend to tune `change_point_prior_scale` and `seasonality_prior_scale` hyperparameters. These two hyperparameters have the most impact on the model.
```
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md]: This is also done btw, forget to mention

File: soda-cl/anomaly-detection.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (baturayo) [reply soda-cl/anomaly-detection.md]: Both of you are right IMO. If I give an example, it might be clearer for the users. However, I also don't want to make it lengthy. So, I'll create a screenshot of an example and add that visualization to the end of the section

File: soda-cl/anomaly-detection.md (hunk 7 above)
Comment (baturayo) [reply soda-cl/anomaly-detection.md]: For the 1st comment, I think you are right @tituskx . People might misunderstand that coverage and mape has different anomaly flagging mechanisms. I'm adjusting that.

File: soda-cl/anomaly-detection.md (hunk 14 above)
Comment (baturayo) [reply soda-cl/anomaly-detection.md:378]: I think `The model tests every possible combination of the listed values for each hyperparameter` has the same meaning as grid search. The person who knows grid search can understand that it is the used method. However, if people don't know what is grid search then they won't understand what we are doing.

File: soda-cl/anomaly-detection.md (hunk 16 above)
Comment (baturayo) [reply soda-cl/anomaly-detection.md:410]: It is true. We will implement the custom warn/fail/buffer values in the second release of AD.

File: soda-cl/anomaly-detection.md (hunk 17 above)
Comment (baturayo) [reply soda-cl/anomaly-detection.md:419]: We already refer to fbprophet parameter tuning section at multiple places. I think people can check the links from there.

**Comment by baturayo** [comment]
Thanks a lot @janet-can and @tituskx for the feedback and the improvements.

File: soda-cl/anomaly-detection.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (tituskx) [reply soda-cl/anomaly-detection.md]: Makes sense!

File: soda-cl/anomaly-detection.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (tituskx) [reply soda-cl/anomaly-detection.md]: It wasn't super clear to me based on the docs that we had done actual tests with hyperparameter tuning, but maybe was a bit too nitpicky. I think it's great right now! 👍

File: soda-cl/anomaly-detection.md (hunk 7 above)
Comment (tituskx) [reply soda-cl/anomaly-detection.md]: > 2. I can, sure. I did a quick search but didn't find an explanation that I loved or was sure was 100% right. Do you have a suggestion for a learn more link?

I didn't find any either 😄, let's just keep it as it is 👍

File: soda-cl/anomaly-detection.md (hunk 8 above)
Comment (tituskx) [reply soda-cl/anomaly-detection.md]: Got it! Agreed 👍

File: soda-cl/anomaly-detection.md (hunk 14 above)
Comment (tituskx) [reply soda-cl/anomaly-detection.md:381]: > I think The model tests every possible combination of the listed values for each hyperparameter has the same meaning as grid search. The person who knows grid search can understand that it is the used method. However, if people don't know what is grid search then they won't understand what we are doing.

Agreed! Reread and agree with both of you that it's clear enough, looks great:)

File: soda-cl/anomaly-detection.md (hunk 16 above)
Comment (tituskx) [reply soda-cl/anomaly-detection.md:413]: got it 🤝

File: soda-cl/anomaly-detection.md (hunk 17 above)
Comment (tituskx) [reply soda-cl/anomaly-detection.md:422]: cool makes sense!

**Review by janet-can — APPROVED** [review]
I adjusted a few tiny bits here and there. LGTM!

**Review by bastienboutonnet — APPROVED** [review]
LGTM. Great documentation work @baturayo 

Here's a "shipit cat" (it's a cat most of the time) but I am at the mercy of the randomisation gods.
